nixos/modules/services/web-servers/nginx/default.nix

```diff
@@ -923,13 +923,11 @@ in
             servers = mkOption {
               type = types.attrsOf (types.submodule {
                 options = {
-                  backup = mkOption {
-                    type = types.bool;
-                    default = false;
-                    description = lib.mdDoc ''
-                      Marks the server as a backup server. It will be passed
-                      requests when the primary servers are unavailable.
-                    '';
```

Comment:
We should keep `backup` and add a `freeformType = attrOf (oneOf [ bool int str ]);` for the other options. Refer to the [documentation](http://nginx.org/en/docs/http/ngx_http_upstream_module.html#server) for what kinds of parameters the `server` directive accepts.

Reply:
Returned `backup` option.
Error with use `freeformType = with types; attrOf (oneOf [ bool int str ]);`:
```
       error: undefined variable 'attrOf'

       at /nix/store/qx77h4vzv2j8lp9s1mvm1xlda8cq45zh-nixos/modules/services/web-servers/nginx/default.nix:925:44:

          924|               type = types.attrsOf (types.submodule {
          925|                 freeformType = with types; attrOf (oneOf [ bool int str ]);
             |                                            ^
          926|                 options = {

```

Reply:
`attrsOf`

Reply:
Thanks, fixed!

Reply:
Sorry, I wasn't clear: I'm suggesting that we get rid of `extraOptions` and instead use the submodule itself to set `server` parameters, using a function like this one:

```nix
let toParameter = key: value:
  if builtins.isBool value
  then lib.optionalString value key
  else "${key}=${toString value}";
```

Then use it to render the `server` directive as follows:

```nix
''
  server ${name} ${concatStringsSep " " (mapAttrsToList toParameter server)};
''
```

The examples would then look something like this:

```nix
{
  servers."backend1.example.com:8080" = { weight = 5; };
  servers."backend2.example.com" = { max_fails = 3; fail_timeout = "30s"; };
  servers."backend3.example.com" = {};
}
```

Reply:
Reworked PR.